### ISSUE TITLE
NaN value for TripletSemiHardLoss when used with TPUs

### ISSUE THREAD

**Issue by sayakpaul** [tensorflow/addons#2408]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow version and how it was installed (source or binary): Binary
- TensorFlow-Addons version and how it was installed (source or binary): Binary
- Python version: 3.7
- Is GPU used? (yes/no): No

**Describe the bug**

I am using `TripletSemiHardLoss` for a project. When I use it with CPUs and GPUs things work as expected. But the moment I switch to TPU-based training the loss goes `NaN`. I am using in-built compile and fit loops so I don't think explicit scaling is required. 

I was suspecting it's probably because of the larger batch size. But I can confirm that's not the case. When using TPUs the batch size I am using is 1024. When I took a sample batch with the size and executed it on a GPU, the loss came out as expected. 

**Code to reproduce the issue**

Unfortunately, I cannot share a code snippet at this point in time. But wanted to know if this behavior is something known.

**Comment by AakashKumarNain** [comment]
Thanks @sayakpaul for reporting this but we need a minimal code example to check the issue. Can you provide a toy example?

**Comment by sayakpaul** [comment]
@AakashKumarNain sure but that will take me at least next week to put it together. I am okay to close it for now and reopen it when I have the toy example. What do you suggest?

**Comment by WindQAQ** [comment]
It's likely to be the known issue of XLA. XLA usually adopts different implementation or even emits different IR on different device for performance issues. This would result in some numeric differences. For example, on tf-nightly, relu with Eigen C++ kernel will propagate NaN while XLA will clip NaN to 0.

https://colab.research.google.com/drive/1b6hTCpudn8m03nQDcXluzHM7ury9Hy2X?usp=sharing

**Comment by sayakpaul** [comment]
That is what I had suspected. Thanks for sharing!

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
